File: GoodsTracker/ringbuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GoodsTracker
{
    class RingBuffer
    {
        const int DEFAULT_BUFFER_SIZE = 64;

        byte[] data  = null;

        int index_producer;
        int index_consumer;
        uint count;

        internal RingBuffer(int size)
        {
            data = new byte[size>0?size: DEFAULT_BUFFER_SIZE];
        }

        internal bool putData(byte ch)
        {
            bool flag = false;

            if (!isFull())
            {

                data[index_producer++] = ch;
                index_producer %= data.Length;
                count++;

                flag = true;
            }

            return flag;
        }

        internal bool getData(out byte ch)
        {
            bool flag = false;

            ch = new byte();

            if (hasData())
            {

                ch = data[index_consumer++];
                index_consumer %= data.Length;
                count--;

                flag = true;
            }

            return flag;
        }

        internal uint getCount()
        {
            return count;
        }

        internal bool isFull()
        {
            return getCount() >= data.Length;
        }

        internal bool hasData()
        {
            return getCount() > 0;
        }

        internal void initBuffer()
        {
            index_consumer  = 0;
            index_producer  = 0;
            count           = 0;

            for (uint i = 0; i < data.Length; i++)
            {
                data[i] = 0;
            }
        }

    }
}

File: platform/esp32-mqtt/components/utils/RingBuffer.c
/*
 * CircularBuffer.c
 *
 *  Created on: 04/06/2017
 *      Author: Jefferson
 */

#include <stdlib.h>

#include "RingBuffer.h"

bool putData(RingBuffer* buffer,char ch){

	bool flag = false;

	if(buffer!=NULL){

	////	EnterCritical();
	//	vPortEnterCritical();

		if(!isFull(buffer)) {

			buffer->data[buffer->index_producer++] = ch;
			buffer->index_producer %= BUFFER_SIZE;
			if(buffer->count<BUFFER_SIZE){
				buffer->count++;
			}

			flag = true;
		}

	////	ExitCritical();
	//	vPortExitCritical();
	}

	return flag;
}
//------------------------------------------------------------------------

bool getData(RingBuffer* buffer,char* ch){

	bool flag = false;

	if(buffer!=NULL){

		////	EnterCritical();
		//vPortEnterCritical();

		if(hasData(buffer)){

			*ch = buffer->data[buffer->index_consumer++];
			buffer->index_consumer %= BUFFER_SIZE;
			if(buffer->count>0){
				buffer->count--;
			}

			flag = true;
		}

		////	ExitCritical();
		//vPortExitCritical();
	}

	return flag;
}
//------------------------------------------------------------------------

inline short getCount(RingBuffer* buffer){

	return buffer!=NULL?buffer->count:0;
}
//------------------------------------------------------------------------

void putString(RingBuffer* buffer,const char* str){

	char* p = (char *)str;

	if(p!=NULL){

		while(*p!='\0'){

			putData(buffer,*p++);
		}
	}
}
//------------------------------------------------------------------------
